File: tools/verify_direction_browser.mjs
import { chromium } from 'playwright';
import { mkdir, writeFile } from 'node:fs/promises';

const outDir = 'tmp/direction-browser';
const directions = {
  '12': [640, 120], '1': [890, 170], '3': [1080, 360], '5': [890, 550],
  '6': [640, 600], '7': [390, 550], '9': [200, 360], '11': [390, 170],
};

async function waitForAtlas(page) {
  await page.waitForFunction(() => !!P && !!G && !!G.on && !!_atlasMask, null, { timeout: 60000 });
}

async function captureCharacter(page, charIndex, name) {
  await page.evaluate((index) => _loadCharAtlas(index), charIndex);
  await page.waitForTimeout(1000);
  await waitForAtlas(page);
  const result = {};
  for (const [clock, [x, y]] of Object.entries(directions)) {
    await page.mouse.move(x, y);
    await page.waitForTimeout(150);
    await page.keyboard.down('KeyD');
    await page.waitForTimeout(100);
    result[clock] = await page.evaluate(() => ({
      facing: P.facing,
      facingDir: _facingDir8(P.facing),
      playerState: P.s,
      charIdx: _charIdx,
      mouseFacing: _mouseFacing,
      gamepad: _gpActive,
    }));
    await page.keyboard.up('KeyD');
    await page.waitForTimeout(100);
    await page.screenshot({ path: `${outDir}/${name}-${clock}.png` });
    const playerScreen = await page.evaluate(() => {
      const rect = C.getBoundingClientRect();
      return {
        x: rect.left + (P.x - G.cam.x + VW / 2) * rect.width / VW,
        y: rect.top + (P.y - G.cam.y + VH / 2) * rect.height / VH,
      };
    });
    await page.screenshot({
      path: `${outDir}/${name}-${clock}-player.png`,
      clip: { x: Math.max(0, playerScreen.x - 96), y: Math.max(0, playerScreen.y - 96), width: 192, height: 192 },
    });
  }
  return result;
}

await mkdir(outDir, { recursive: true });
const browser = await chromium.launch({ headless: true });
const page = await browser.newPage({ viewport: { width: 1280, height: 720 } });
const consoleErrors = [];
page.on('console', (message) => {
  if (message.type() === 'error') consoleErrors.push(message.text());
});
await page.goto('http://127.0.0.1:3333/game.html?test=1&testchar=1&slot=direction-browser', { waitUntil: 'networkidle' });
await page.waitForTimeout(5000);
console.log('BOOT_STATE', JSON.stringify(await page.evaluate(() => ({
  p: typeof P,
  playerReady: !!P,
  g: typeof G,
  atlas: typeof _atlasMask,
  atlasReady: !!_atlasMask,
  boot: document.getElementById('bootLoading')?.textContent,
  canvases: document.querySelectorAll('canvas').length,
}))));
await waitForAtlas(page);
const result = {
  warrior: await captureCharacter(page, 0, 'warrior'),
  silvertail: await captureCharacter(page, 1, 'silvertail'),
  consoleErrors,
};
await writeFile(`${outDir}/result.json`, JSON.stringify(result, null, 2));
console.log(JSON.stringify(result, null, 2));
await browser.close();

File: tools/verify_direction_browser.py
from pathlib import Path
from playwright.sync_api import sync_playwright

OUT = Path('tmp/direction-browser')
OUT.mkdir(parents=True, exist_ok=True)

DIRECTIONS = {
    '12': (640, 120), '1': (890, 170), '3': (1080, 360), '5': (890, 550),
    '6': (640, 600), '7': (390, 550), '9': (200, 360), '11': (390, 170),
}

def wait_for_atlas(page):
    page.wait_for_function("() => typeof P !== 'undefined' && typeof G !== 'undefined' && _atlasMask && _atlasMask.complete")

def capture_character(page, char_index, name):
    page.evaluate("idx => _loadCharAtlas(idx)", char_index)
    page.wait_for_timeout(1000)
    wait_for_atlas(page)
    results = {}
    for clock, (x, y) in DIRECTIONS.items():
        page.mouse.move(x, y)
        page.wait_for_timeout(150)
        page.keyboard.down('KeyD')
        page.wait_for_timeout(100)
        state = page.evaluate("""() => ({
            facing: P.facing,
            facingDir: _facingDir8(P.facing),
            state: P.s,
            charIdx: _charIdx,
            mouseFacing: _mouseFacing,
            gamepad: _gpActive
        })""")
        page.keyboard.up('KeyD')
        page.wait_for_timeout(100)
        page.screenshot(path=str(OUT / f'{name}-{clock}.png'))
        results[clock] = state
    return results

with sync_playwright() as p:
    browser = p.chromium.launch(headless=True)
    page = browser.new_page(viewport={'width': 1280, 'height': 720})
    logs = []
    page.on('console', lambda msg: logs.append(f'{msg.type}: {msg.text}'))
    page.goto('http://127.0.0.1:3333/game.html?test=1&testchar=1&slot=direction-browser', wait_until='networkidle')
    wait_for_atlas(page)
    result = {
        'warrior': capture_character(page, 0, 'warrior'),
        'silvertail': capture_character(page, 1, 'silvertail'),
        'consoleErrors': [line for line in logs if line.startswith('error:')],
    }
    (OUT / 'result.json').write_text(__import__('json').dumps(result, indent=2), encoding='utf-8')
    print(__import__('json').dumps(result, indent=2))
    browser.close()
